### ISSUE TITLE
Better support/documentation for implementing iterative algorithms

### ISSUE THREAD

**Issue by gabrielgrant** [pachyderm/pachyderm#2999]
**What is the goal / desired outcome?**

Pachyderm doesn't currently have a well-trodden path to implement an iterative algorithm, where a result is recursively refined until it reaches some acceptable end state. This end state may be defined by total number of iterations, but is more likely defined as an acceptable convergence point either determined by some measure of absolute accuracy, or by comparing the current iteration's result to that of previous iterations and observing that the estimate is no longer (significantly) improving.

The specific example that Drayton Munster (@dwmunster in #users) is trying to work through is parallelizing multiple MCMC runs (something like http://dfm.io/emcee/current/), but the same pattern basically applies to most optimization algorithms (Genetic Algorithms, simulated annealing etc.) / similarly-formulated physics simulations (protein folding, graph layout, etc.):

- choose a set of start points (possibly parallel)
- sample your function at those points (parallel)
- pick the "good" ones (possibly parallel)
- determine if your results are "good enough" (likely not parallel, though _ideally_ this would be able to use an iterative/parallel reduce, if pach support existed)
  - if so, stop
  - if not, choose a new set of points to evaluate (likely based on the best candidates from the previous iteration) and repeat

**What is the ask?**

Ideally: a way to achieve this within the pachyderm execution model, that preserves provenance across iterations and doesn't require hacks.

At the least: a clearly-defined recommendation for how to achieve this pattern using the pachyderm primitives available today.

**If there is a way to accomplish this today via workaround, what does that require?**

It seems the most straightforward way to implement this today would be to have a "DAG" that looks like this:

```
[iterations] --> [samples] --> [stop checks]
    ^                                   .
    .                                   .
    . . . . . . . . . . . . . . . . . . .
```

Processing begins by adding a new/first iteration (containing initial conditions). This generates a set of points at which to sample the objective function. The samples are evaluated to determine whether the stop condition has been met. If so, final results get written out; if not, a new iteration, with refined initial conditions, gets added to the `iterations` repo (dotted line represents this recursive pseudo-link -- since data is directly committed to the upstream repo, rather than passed as output, the provenance tracking chain gets broken here)

One way I can think of to get full provenance tracking would be to dynamically "unroll" the iteration loop: if the evaluation step determined that another iteration was needed, it could create another downstream step in the DAG. Actually creating a whole chain of pipelines running the same code seems pretty absurd, so I wouldn't recommend implementing this except _maybe_ if the provenance tracking were truly critical, and the expected number of iterations were (very) small. But I could imagine some sort of model baked into Pachyderm that logically implemented cyclical provenance tracking in this manner.

In @dwmunster's case, the final evaluation (potentially?) requires all samples from all iterations be present. Because of that, it seems the most straightforward setup would be to have each iteration's samples be written to a new directory in the samples repo. At some point (tens of thousands of iterations over tens of thousands of sample points) having all these files in a single commit will start to cause significant overhead for each iteration. The upper bound on the total number of files in a commit should be significantly increased by #2984 but there is still some cost to transferring this extra metadata to workers when doing processing. There's also a cost to checking that previous iterations don't need to be re-evaluated. I _think_ there may be a way to divide the files into datums in such a way that the cost would be relative to the number of iterations (1000s), not the total number of evaluations (potentially 10s of millions), which wouldn't be so significant, but the naive pattern of having `iterations/<iteration-id>/<stample-id>` and taking `/iterations/*/*` as input to the sampling step would require checking every one of those past evaluations, which would likely become prohibitively expensive. This cost may be mitigated by @brycemcanally 's current work (#2986 ?), and would likely be eliminated by #2993 

@dwmunster was considering writing each iteration to a separate branch to avoid these performance concerns, but that would require complications of having to update downstream pipelines to take different branches (potential use-case for #2882), so is going to try the above pattern first.

For many (most?) other cases, such Genetic Algorithms, there's a bit of a difference in that you're generally only concerned with the surviving population at a given iteration, so although it has the same general iteration pattern, the working dataset doesn't need to continue to grow the way it sounds like this MCMC case will, so the samples can likely be overwritten, mitigating the above-mentioned concerns.

**Comment by gabrielgrant** [comment]
Relevant conversation: https://pachyderm-users.slack.com/archives/C3UN0S0Q0/p1528926671000463

<details>
  <summary>screenshotted</summary><p>

![image](https://user-images.githubusercontent.com/493968/41389207-4999393c-6f44-11e8-90f6-8d5a36a6d879.png)
  </p>
</details>

<details>
  <summary>searchable version</summary><p>
Drayton Munster [2:51 PM]
Are there any examples/suggestions on building an optimization/analysis loop with Pachyderm?


Gabriel Grant [2:53 PM]
@dwmunster pachyderm can certainly be useful for that sort of thing. mind explaining a bit more specifically about your use-case?
are you talking specifically about implementing an iterative optimization algorithm? or just about performing optimization of, say, an NN's hyperparameters? (edited)

Drayton Munster [2:55 PM]
The specific example I had in mind was using Pachyderm to handle the parallelization and execution for MCMC sampling.
My current thought was to have the “analysis” pipeline take, e.g., a UUID and the MCMC code. Each iteration would then be pushed to an “execution” pipeline with branches keyed by ID and iteration number.

Drayton Munster [3:09 PM]
uploaded this image: example.png 

Drayton Munster [3:11 PM]
[Apparently you can’t edit comments on uploaded images, so reproduced here in comment form.]
Specifically, I’d be interested in using Pachyderm to manage the execution/provenance of the colored block. Each P_i may have many hundreds/thousands of sample points, each of which can be evaluated independently.

Gabriel Grant [3:20 PM]
ok, full disclosure: i have fairly rudimentary understanding of MCMC (i've only used it a little in a non-parallelized context -- my understanding is that in the naive case it is an iterative algo where determining sample point n+1 depends on evaluation of sample point n), but if the samples can be independently computed, that sounds like a good candidate for parallelization (edited)

Drayton Munster [3:22 PM]
Choosing the next batch of sample points can be complicated, but that general idea (iteration batch n+1 depends on iteration batch n) is the important bit
And is the same for MCMC or iterative optimization algorithms

Gabriel Grant [3:23 PM]
but each iteration of P_i is evaluating a whole set of independent points, right?

Drayton Munster [3:24 PM]
Yes

Gabriel Grant [3:27 PM]
ok, so that "pure map" type function is certainly well within pachyderm's wheelhouse. the "outer loop" iteration of evaluating whether your set of evaluated points is sufficient to reach a termination condition, or if it should trigger another iteration, is a bit more tricky, but certainly doable (i'd probably just have an evaluation pipeline that conditionally creates a new commit to an upstream ingress repo if another iteration is needed) (edited)

Drayton Munster [3:28 PM]
That's the idea I was trying to describe earlier, wanted to make sure I wasn't missing a better way.

Gabriel Grant [3:31 PM]
ah, ok -- sorry, just needed a bit more context to be able to follow what you were asking. i'm still not sure i totally follow what you mean about having branches keyed by ID and iteration number, though
my thinking was that each iteration would basically be triggered by a commit on an input repo of a list of points at which to sample
each of those sample points would become a datum, which could be executed in parallel
then you'd presumably need some sort of a reducer to determine whether to iterate again

Drayton Munster [3:34 PM]
The ID and iteration branches would just be a slightly more readable way to track the execution.

Gabriel Grant [3:34 PM]
(and something to generate the next set of points to evaluate)
ok, yea, what i'm describing is certainly a pretty naive way of doing this
i'm not sure i totally follow how the ID and iteration branches would work exactly, though. mind spelling that out a bit more for me?

Drayton Munster [3:38 PM]
The final result may depend on the entire history (e.g. all the P_i and y_i). The way you described requires keeping a list of iterations to commit ids.

Gabriel Grant [3:38 PM]
ah, i see, so at the end you'd have to reduce over all the commits to get a final result?

Drayton Munster [3:38 PM]
Yes

Gabriel Grant [3:38 PM]
gotcha

Drayton Munster [3:39 PM]
So ID plus iteration is just a human friendly way to track in case manual inspection is necessary
For an optimization example, consider making a convergence plot

Gabriel Grant [3:40 PM]
i wonder whether it would make sense to just keep all the history in the head commit, then, and just add a new directory for each iteration's results?
(meaning each commit would just append another directory, and your final evaluation would see all results in a single commit to reduce over)
do you have any idea roughly how many iterations we're talking about?

Drayton Munster [3:44 PM]
Optimization would like be small (<100), but MCMC could be many thousands of loop iterations

Gabriel Grant [3:45 PM]
sorry, just to make sure i'm clear: you're saying the number of outer loop iterations would be <100, but the number of samples evaluated within each outer loop run (ie within each P_i step) would be many thousands?

Drayton Munster [3:46 PM]
Both the number of independent points and the number of outer iterations could be large

Gabriel Grant [3:47 PM]
ah, ok
and what does this reduce/looper step that determines whether to run another outer iteration (and computes which points need to be evaluated?) look like?

Drayton Munster [3:49 PM]
For example: http://dfm.io/emcee/current/

Gabriel Grant [3:49 PM]
does it need to see all the historical data every time it runs?

Drayton Munster [3:49 PM]
The basic usage example has 100 independent points with 1000 outer iterations
Determining if another iteration is necessary could, in principle, require the full history

Gabriel Grant [3:52 PM]
ok, thanks, that's useful to have something concrete to look at (sorry again for the elementary MCMC questions)

Drayton Munster [3:53 PM]
But in these examples, they just need the most recent and the only reduction at the end needs the full history
For example, we might want the most likely point (maximum) out of the 100 x 1000 evaluations

Gabriel Grant [3:56 PM]
ah, i see, looks like this multiple "walker" concept is what i was missing from my understanding of MCMC: http://dfm.io/emcee/current/user/faq/#what-are-walkers
gotcha

Drayton Munster [3:56 PM]
I don't mind doing the iteration calculation and reduction in a single container, but the parallel walker eval is what I'm interested in

Gabriel Grant [3:57 PM]
so, at some point you're going to need to see everything

Drayton Munster [3:57 PM]
Not all MCMC algorithms have the multiple walkers, but that one does. For an optimization example, consider evolutionary algorithms or simulated annealing

Gabriel Grant [4:06 PM]
ok, well yes, parallelizing over walkers sounds like it should certainly be doable. if your reduce step is going to run over all the results, my inclination would be to model this as an append-only workload where each datum is a point at which to evaluate, and subsequent iterations just get added as additional directories (mostly for the sake of making it easier to consume the results together). but having one branch per iteration would basically be equivalent (a bit more work to set up, but would save some overhead of having a ton of mostly-unused files in the commit tree) (edited)

Drayton Munster [4:33 PM]
How expensive is that overhead for the unused files? Keeping the iterations together in different folders is tempting.

Gabriel Grant [4:41 PM]
basically the main cost is that there's some extra metadata transferred to workers when doing processing, but that should only be paid once per job (at most; i believe it should generally get cached). there also, historically, has been an upper bound on the total number of files in a commit because of the requirement that said metadata be loaded into memory when finalizing a commit. but there's work going on at this very moment that should be landing soon to remove that limitation: https://github.com/pachyderm/pachyderm/pull/2984 :slightly_smiling_face: (edited)
GitHub
Bolt hashtree by jdoliner · Pull Request #2984 · pachyderm/pachyderm

Drayton Munster [4:43 PM]
Nice, I’ll see if I can put together a (fairly) minimal example. Thanks for the help

Gabriel Grant [7:33 PM]
@dwmunster should have mentioned: there'd also be a slight cost to recognize that previous iterations don't need to be re-evaluated, but i believe the pattern we're describing could be implemented in such a way that that cost would be relative to the number of iterations (1000s), not the total number of evaluations (100s of 1000s), so shouldn't be too significant (it's basically one network request per check)
we do also have a proposal to remove that overhead completely: https://github.com/pachyderm/pachyderm/issues/2993
GitHub
Propagate list of changed files through PPS DAG · Issue #2993 · pachyderm/pachyderm
Background Right now, PPS jobs avoid the cost of reprocessing datums by this mechanism: Each worker claims some subset of the job&#39;s datums The worker hashes each datum and checks if the hash i...
re: GAs -- there's a bit of a difference in that you're generally only concerned with the surviving population at a given iteration, so the working dataset doesn't continue to grow the way it sounds like your MCMC one will, but it is the same general iteration pattern, yea
glad i could help confirm that you were basically on the right track. look forward to seeing what you come up with, and let us know if there's anything else we can do to help. wrote up a summary/proposal to make what we're talking about easier here: https://github.com/pachyderm/pachyderm/issues/2999
GitHub
Better support/documentation for implementing iterative algorithms · Issue #2999 · pachyderm/pachyderm
  </p>
</details>

**Comment by JoeyZwicker** [comment]
Isn’t this what incremental is for? “I need to look at my result from last time to decide what to do this time”

**Comment by gabrielgrant** [comment]
The issues are:
1. the incrementality here spans multiple pipeline stages
2. the triggering of subsequent iteration is caused by the output of a previous run, rather than by some externally-initiated commit

neither of these are well-handled by our current notion of incremental processing

**Comment by gabrielgrant** [comment]
To expand on what I mentioned in standup today: one of the cases where this pattern is perhaps most relevant our core use-case is for hyper-parameter tuning of ML models. Right now we do a good job of supporting manual hyper-parameter tuning (where hyper-parameters live in an ingress repo and are manually updated) and dumb grid-search sweeps (where the various hyper-parameter options to test are predefined, they can just be put in an input repo and each become a datum that triggers a training run) but any more intelligent/automated hyper-parameter tuning requires the type of feedback loop that this issue discusses

Especially for something like random search, where (as the name suggests) the detailed decisions/performance of the final model will depend on the specific random numbers chosen at each iteration of the tuning, having provenance tracking seems like it could be particularly important.

**Comment by jdoliner** [comment]
This is a decidedly bad use case for Pachyderm right now and some major changes would need to happen before we could really have first class support for this. The crux of the problem is that an iterative model plays very poorly with our provenance system which works by laying out all the downstream commits and then filling in the results as they become available. In the case of an iterative algorithm this set of commits is infinite, the processing shouldn't actually be infinite, although if the algorithm fails to converge for some reason it could be. But even if it doesn't our provence system doesn't have any way to know at which point things will converge. To represent this in a first class way we'll likely need to have some sort of lazily computed provenance system that can compute things symbolically without actually writing out all of the commits.

In terms of a work around I think the method you describe here is as good as it's going to get. Basically manually ferrying data back to the top of the dag when you want to trigger another iteration. This makes it explicit when you consider the operation to be done, whenever you stop ferrying data back, and can also be done quite performantly using the `CopyFile` method.

**Comment by gabrielgrant** [comment]
Yea, I actually don't think this workaround is that bad, if implemented with the output of the last "logical" pipeline stage basically being (instructions for?) what gets inserted back at the top of the DAG, then having an "egress pipeline" that just does the `CopyFile` operation

While some type of lazily computed notion of infinite provenance could probably be baked up to solve this, the type of first-class support I was thinking could most easily exist would basically be (logically) equivalent to dynamically extending the DAG downwards after each iteration, but without actually creating new k8s resources for the downstream pipelines (to avoid the excessive computation resource cost). My understanding is that basically would work today within our model, just wouldn't be recommended because:

1. it's prohibitively expensive to run many instances of the same pipeline(s) (and maybe not even, with aggressive spin-down)
2. it would be pretty ugly to look at all those extra DAG elements
3. we don't often see thousands of DAG elements, so we probably have some other unknown scaling constraints

Re: 1 -- It seems like not recreating the duplicate downstream pipelines is a conceivable optimization to make (and one we've discussed in the context of pipeline branching (#1036) -- I think this would basically entail noticing that multiple pipelines use the same docker image

Re: 2 -- allowing those workers to be reused, and making it so (potentially large numbers) of related DAG elements don't pollute your view of the system is what we're after with the DAG organization work

Re: 3 -- removing these limitations may not be a high priority, but doing so certainly wouldn't be fundamentally outside our model

The fundamentally new piece that _would_ be needed to have first-class support would be some notion of branching: basically a way for a pipeline to tell pach that it's output should trigger the (logical) creation of a new "iteration" of a section of the DAG that consumes it's output. But that logic could also easily just be wrapped up in a "branching condition pipeline" that made the DAG modifications

That being said, there would obviously be a few benefits to having it baked in:

- this could likely be stored more efficiently as some sort of "instance count", rather than actually duplicating all the elements in etcd, which would start to matter at some point
- having a notion of updating all the copied pipelines together

But those seem like they could pretty easily be worked around today, it likely doesn't make sense to bake it in that unless we see use of this workaround cropping up a lot and start having people run into these issues
